### ISSUE TITLE
Problem with the HUD in Gauntlet Legends

### ISSUE THREAD

**Issue by LegendOfDragoon** [purplemarshmallow/z64#62]
@purplemarshmallow do you know which commit(s) got the HUD to even display? Nice work. I would like to find out so I can port it to other plugins like Glide64 if possible.

![image](https://user-images.githubusercontent.com/10492645/28243091-7320d404-6973-11e7-9256-4cd48ac470b1.png)

Angrylion's commit r96 fixed this issue, I believe. https://sourceforge.net/p/angrylions-stuff/code/96/tree/trunk/mylittle-nocomment/n64video.cpp?diff=55bb05dc27184618d775bec1:95&diformat=regular

**Comment by LegendOfDragoon** [comment]
Seems like https://github.com/purplemarshmallow/z64/commit/44332ac1f4a23f6ce68d49ad3471e0b1c58ef238 was the commit to get the HUD displaying.

**Comment by purplemarshmallow** [comment]
Nice find!

I used Glide64 as reference for the commit you linked. So if the HUD does not show up with Glide64 it must be a different problem

**Comment by LegendOfDragoon** [comment]
I dug deeper and believe it's a blending issue.

At the beggining of the game (where I took screenshot), the location of the HUD in RDRAM starts at 0x802B07F0. The RDP loads that data into TMEM.

It is in this function call where the problem happens https://github.com/purplemarshmallow/z64/blob/master/src/rgl_tiles.cpp#L262

You can set a breakpoint by checking if address == 0x002B07F0. I believe the total size of the HUD is somewhere around 0x7000 or 0x8000 bytes. It's constantly overwriting that area, when it shouldn't.

**Comment by purplemarshmallow** [comment]
What exactly are you investigating? Black bars in the HUD?

**Comment by LegendOfDragoon** [comment]
>Black bars in the HUD?

Yes.

**Comment by purplemarshmallow** [comment]
>It is in this function call where the problem happens https://github.com/purplemarshmallow/z64/blob/master/src/rgl_tiles.cpp#L262

You have the hires fb setting disabled? With hires fb enabled it should never reach this line. If hires fb is disabled it writes frame buffers to rdram. This does not always work well. I found hires fb enabled is usually better/more compatible

**Comment by LegendOfDragoon** [comment]
>You have the hires fb setting disabled?

Yes. You're right about it never reaching that line if it's on. I still have the issue though, although it looks a little different. It's black mini squares instead of the black bar. I wonder how to debug it then.

Edit: 1 problem with hi-res framebuffer is that the HUD won't be loaded after making a save state with z64gl and loading the save state later on.

**Comment by purplemarshmallow** [comment]
Maybe the graphical debugger is useful here. You can press P to pause and then click objects to view details. But mouse coordinates are a bit buggy. With +/- you can view different frame buffers. I can't test it right now but maybe clicking the black bars will give a clue

>Edit: 1 problem with hi-res framebuffer is that the HUD won't be loaded after making a save state with z64gl and loading the save state later on.

That's true that's a problem with hardware framebuffer. For debugging it might be best to load a savestate made with angrylions plugin

**Comment by purplemarshmallow** [comment]
I got the HUD to show up correctly when loading a savestate made with angrylion's plugin. That should give a hint where the problem lies. The HUD is a framebuffer texture. It can load correctly but the plugin renders this texture incorrectly
![00000004](https://user-images.githubusercontent.com/7283660/29486960-42c97b04-84f2-11e7-8db0-24bdcfa730d3.jpg)

**Comment by purplemarshmallow** [comment]
So I view how the hud is rendered in the debugger but I'm not sure what's wrong and what's going on exactly. I fear my knowledge ends here

![4](https://user-images.githubusercontent.com/7283660/29487183-854ee0ba-84f5-11e7-82ec-09e4d5e370e9.jpg)
![5](https://user-images.githubusercontent.com/7283660/29487182-854ca76e-84f5-11e7-8003-611b57f3cc3b.jpg)
![6](https://user-images.githubusercontent.com/7283660/29487181-8548b2b2-84f5-11e7-93e1-b8d3c20209a9.jpg)

**Comment by cxd4** [comment]
I'd be happy enough if the combiner issue with the black geometry throughout the entire game was fixed.

I doubt it's linked to the GLSL issue I fixed a few PRs back, but I can't test for sure because I don't have GNU make installed to WINE and have no way to build the latest version of the plugin from source for Windows.  A new release with the geometry colored in would be great.

**Comment by purplemarshmallow** [comment]
Yes the black geometry is the more serious problem. I assume @LegendOfDragoon wanted to investigate the Hud instead because it affects other plugins as well.

>I doubt it's linked to the GLSL issue I fixed a few PRs back,

Looks the same before and after

**Comment by LegendOfDragoon** [comment]
I'll have to try the debugger then. I hope i can learn how to effectively use it. When using hi-res FB, it kind of looks similar to the issue Angrylion's used to have.

>I assume @LegendOfDragoon wanted to investigate the Hud instead because it affects other plugins as well.

That is indeed correct.

**Comment by cxd4** [comment]
I actually just edited z64gl.dll in a hex editor to fix the GLSL strings for LOD fraction to fix the error that way anyway.  It's interesting because you don't get the error on Windows...it's an error on Linux, nothing at all on Windows, and a warning inside WINE on Linux running Windows. :D 

The black geometry is present in other plugins as well.  I'm not sure, but IIRC Glide64 and RiceVideo are probably the only ones to have hacked the ucode in right to not have the black geometry.

**Comment by purplemarshmallow** [comment]
That's weird. Yes I get no Glsl error.

If you fix the errors does the geometry appear or is it still black?

**Comment by cxd4** [comment]
Oh nah.  It fixes graphics issues on Linux, but in WINE it's only a GLSL "warning" (not an error); it only removes the console spamming so I have a nice clean console with space for the real error messages I'd care about.  And then there's native Windows non-WINE where there's no warning or error.

As far as debugging the black geometry issue, it does at least work in the software, non-GL z64.dll.

**Comment by LegendOfDragoon** [comment]
They recently fixed some issues in ParaLLEl https://github.com/libretro/parallel-n64/commit/505c0f2a7dd443260d9d3bd4351934548414f731. I checked z64gl's source and it's missing the key_color stuff. Although their HUD is still bugged.

Edit: I recently tested Mischief Makers and the rocks are missing in the intro. So i think there's a good chance both the HUD and missing rocks are caused by the same issue. I'd fix it myself if I understood z64gl's code and OpenGL better.
